In [14]:
import pandas as pd
import numpy as np
# Load the BTC dataset
column_names = ['timestamp', 'open', 'high', 'low', 'close', 'volume', 'close_time', 'quote_asset_volume', 'number_of_trades', 'taker_buy_base_asset_volume', 'taker_buy_quote_asset_volume', 'ignore']

df = pd.read_csv('/content/BNBUSDT (1).csv', header=None, names=column_names)

# Display the first few rows of the data
df.head()

/tmp/ipython-input-14-1855742643.py:6: DtypeWarning: Columns (1,2,3,4,5,7,8,9,10,11) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/content/BNBUSDT (1).csv', header=None, names=column_names)


timestamp  open  high  low  close  volume  \
0            timestamp  open  high  low  close  volume   
1  2017-11-06 03:54:00   1.5   1.5  1.5    1.5   10.83   
2  2017-11-06 03:55:00   1.3   1.3  1.3    1.3     1.0   
3  2017-11-06 03:56:00   1.3   1.3  0.5    0.5    19.0   
4  2017-11-06 03:57:00   0.5  0.61  0.5   0.61   253.0   

                close_time  quote_asset_volume  number_of_trades  \
0               close_time  quote_asset_volume  number_of_trades   
1  2017-11-06 03:54:59.999              16.245                 2   
2  2017-11-06 03:55:59.999                 1.3                 1   
3  2017-11-06 03:56:59.999              24.356                 6   
4  2017-11-06 03:57:59.999              151.66                17   

   taker_buy_base_asset_volume  taker_buy_quote_asset_volume  ignore  
0  taker_buy_base_asset_volume  taker_buy_quote_asset_volume  ignore  
1                        10.83                        16.245       0  
2                          0.0                           0.0       0  
3                         0.33                         0.429       0  
4                          0.0                           0.0       0

In [15]:
# Check data types
print("Data types of each column:")
print(df.dtypes)

Data types of each column:
timestamp                       object
open                            object
high                            object
low                             object
close                           object
volume                          object
close_time                      object
quote_asset_volume              object
number_of_trades                object
taker_buy_base_asset_volume     object
taker_buy_quote_asset_volume    object
ignore                          object
dtype: object


In [16]:
# Step 1: Identify columns to convert
numeric_columns = ['open', 'high', 'low', 'close', 'volume',
                   'quote_asset_volume', 'number_of_trades',
                   'taker_buy_base_asset_volume', 'taker_buy_quote_asset_volume']

# Step 2: Convert object columns to numeric
for col in numeric_columns:
    df[col] = pd.to_numeric(df[col], errors='coerce')

# Step 3: Check for NaN values
print("NaN values in each column after conversion:")
print(df.isna().sum())

# Step 4: Handle NaN values (example: fill with 0 or drop rows)
df.fillna(0, inplace=True)
# Alternatively, to drop rows with NaN values
# df.dropna(inplace=True)

# Verify the data types after conversion
print("Data types after conversion:")
print(df.dtypes)

NaN values in each column after conversion:
timestamp                       0
open                            1
high                            1
low                             1
close                           1
volume                          1
close_time                      0
quote_asset_volume              1
number_of_trades                1
taker_buy_base_asset_volume     1
taker_buy_quote_asset_volume    1
ignore                          0
dtype: int64
Data types after conversion:
timestamp                        object
open                            float64
high                            float64
low                             float64
close                           float64
volume                          float64
close_time                       object
quote_asset_volume              float64
number_of_trades                float64
taker_buy_base_asset_volume     float64
taker_buy_quote_asset_volume    float64
ignore                           object
dtype: object


In [17]:
# Display the columns of the DataFrame
print(df.columns)

# Display the first few rows of the DataFrame
print(df.head())

Index(['timestamp', 'open', 'high', 'low', 'close', 'volume', 'close_time',
       'quote_asset_volume', 'number_of_trades', 'taker_buy_base_asset_volume',
       'taker_buy_quote_asset_volume', 'ignore'],
      dtype='object')
             timestamp  open  high  low  close  volume  \
0            timestamp   0.0  0.00  0.0   0.00    0.00   
1  2017-11-06 03:54:00   1.5  1.50  1.5   1.50   10.83   
2  2017-11-06 03:55:00   1.3  1.30  1.3   1.30    1.00   
3  2017-11-06 03:56:00   1.3  1.30  0.5   0.50   19.00   
4  2017-11-06 03:57:00   0.5  0.61  0.5   0.61  253.00   

                close_time  quote_asset_volume  number_of_trades  \
0               close_time               0.000               0.0   
1  2017-11-06 03:54:59.999              16.245               2.0   
2  2017-11-06 03:55:59.999               1.300               1.0   
3  2017-11-06 03:56:59.999              24.356               6.0   
4  2017-11-06 03:57:59.999             151.660              17.0   

   taker_buy_b

In [18]:
# Display rows with NaN values
print(df[df.isna().any(axis=1)])

Empty DataFrame
Columns: [timestamp, open, high, low, close, volume, close_time, quote_asset_volume, number_of_trades, taker_buy_base_asset_volume, taker_buy_quote_asset_volume, ignore]
Index: []


In [19]:
# Drop rows with any NaN values
df_cleaned = df.dropna()

In [20]:
# Check the first few values in the timestamp column
print(df['timestamp'].head())

0              timestamp
1    2017-11-06 03:54:00
2    2017-11-06 03:55:00
3    2017-11-06 03:56:00
4    2017-11-06 03:57:00
Name: timestamp, dtype: object



Create a new DataFrame with the first 200,000 rows of `df` and display its head and shape to verify.



In [22]:
df_subset = df.tail(200000).copy()
print("First few rows of the subset DataFrame:")
display(df_subset.head())
print("\nShape of the subset DataFrame:")
print(df_subset.shape)

First few rows of the subset DataFrame:


timestamp    open   high     low   close   volume  \
3721496  2024-12-09 05:42:00  724.62  725.0  724.62  725.00   72.711   
3721497  2024-12-09 05:43:00  725.00  725.4  724.80  725.33  308.538   
3721498  2024-12-09 05:44:00  725.34  725.9  725.33  725.86  108.865   
3721499  2024-12-09 05:45:00  725.86  726.4  725.86  726.00  171.246   
3721500  2024-12-09 05:46:00  726.01  726.5  725.90  726.50  171.282   

                         close_time  quote_asset_volume  number_of_trades  \
3721496  2024-12-09 05:42:59.999000         52703.81100             584.0   
3721497  2024-12-09 05:43:59.999000        223711.91979             691.0   
3721498  2024-12-09 05:44:59.999000         78988.34144             669.0   
3721499  2024-12-09 05:45:59.999000        124344.51358             847.0   
3721500  2024-12-09 05:46:59.999000        124401.11192             796.0   

         taker_buy_base_asset_volume  taker_buy_quote_asset_volume ignore  
3721496                       47.147                   34174.13693      0  
3721497                      228.144                  165425.09687      0  
3721498                       86.450                   62723.76861      0  
3721499                      132.689                   96344.56601      0  
3721500                      125.605                   91227.12815      0


Shape of the subset DataFrame:
(200000, 12)



Convert the 'timestamp' column to datetime objects, set it as the index, and select the 'close' column.



In [23]:
df_subset = df_subset.iloc[1:].copy()
df_subset['timestamp'] = pd.to_datetime(df_subset['timestamp'])
df_subset.set_index('timestamp', inplace=True)
close_series = df_subset['close']

print("DataFrame with timestamp index:")
display(df_subset.head())
print("\nSelected 'close' series:")
display(close_series.head())

DataFrame with timestamp index:


open    high     low   close   volume  \
timestamp                                                      
2024-12-09 05:43:00  725.00  725.40  724.80  725.33  308.538   
2024-12-09 05:44:00  725.34  725.90  725.33  725.86  108.865   
2024-12-09 05:45:00  725.86  726.40  725.86  726.00  171.246   
2024-12-09 05:46:00  726.01  726.50  725.90  726.50  171.282   
2024-12-09 05:47:00  726.49  726.62  726.30  726.62  214.000   

                                     close_time  quote_asset_volume  \
timestamp                                                             
2024-12-09 05:43:00  2024-12-09 05:43:59.999000        223711.91979   
2024-12-09 05:44:00  2024-12-09 05:44:59.999000         78988.34144   
2024-12-09 05:45:00  2024-12-09 05:45:59.999000        124344.51358   
2024-12-09 05:46:00  2024-12-09 05:46:59.999000        124401.11192   
2024-12-09 05:47:00  2024-12-09 05:47:59.999000        155460.92414   

                     number_of_trades  taker_buy_base_asset_volume  \
timestamp                                                            
2024-12-09 05:43:00             691.0                      228.144   
2024-12-09 05:44:00             669.0                       86.450   
2024-12-09 05:45:00             847.0                      132.689   
2024-12-09 05:46:00             796.0                      125.605   
2024-12-09 05:47:00            1064.0                      109.814   

                     taker_buy_quote_asset_volume ignore  
timestamp                                                 
2024-12-09 05:43:00                  165425.09687      0  
2024-12-09 05:44:00                   62723.76861      0  
2024-12-09 05:45:00                   96344.56601      0  
2024-12-09 05:46:00                   91227.12815      0  
2024-12-09 05:47:00                   79773.38015      0


Selected 'close' series:


timestamp
2024-12-09 05:43:00    725.33
2024-12-09 05:44:00    725.86
2024-12-09 05:45:00    726.00
2024-12-09 05:46:00    726.50
2024-12-09 05:47:00    726.62
Name: close, dtype: float64


Import the necessary class and fit the ARIMA model to the time series data.



In [24]:
from statsmodels.tsa.arima.model import ARIMA

# Instantiate and fit the ARIMA model
model = ARIMA(close_series, order=(5,1,0))
model_fit = model.fit()

# Display the model summary
print(model_fit.summary())

/usr/local/lib/python3.11/dist-packages/statsmodels/tsa/base/tsa_model.py:473: ValueWarning: A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.
  self._init_dates(dates, freq)
/usr/local/lib/python3.11/dist-packages/statsmodels/tsa/base/tsa_model.py:473: ValueWarning: A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.
  self._init_dates(dates, freq)
/usr/local/lib/python3.11/dist-packages/statsmodels/tsa/base/tsa_model.py:473: ValueWarning: A date index has been provided, but it has no associated frequency information and so will be ignored when e.g. forecasting.
  self._init_dates(dates, freq)


                               SARIMAX Results                                
Dep. Variable:                  close   No. Observations:               199999
Model:                 ARIMA(5, 1, 0)   Log Likelihood             -156541.071
Date:                Wed, 30 Jul 2025   AIC                         313094.143
Time:                        06:27:14   BIC                         313155.379
Sample:                             0   HQIC                        313112.166
                             - 199999                                         
Covariance Type:                  opg                                         
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
ar.L1          0.0105      0.000     26.530      0.000       0.010       0.011
ar.L2         -0.0004      0.001     -0.686      0.492      -0.001       0.001
ar.L3         -0.0031      0.001     -4.228      0.0


Generate predictions using the fitted ARIMA model for the next 10 steps.



In [25]:
# Generate predictions for the next 10 steps
n_future_steps = 10
predictions = model_fit.predict(start=len(close_series), end=len(close_series) + n_future_steps - 1)

# Print the first few predictions
print("Generated predictions:")
print(predictions)

Generated predictions:
199999    602.310400
200000    602.310811
200001    602.308207
200002    602.310056
200003    602.311201
200004    602.311214
200005    602.311214
200006    602.311235
200007    602.311207
200008    602.311193
Name: predicted_mean, dtype: float64


/usr/local/lib/python3.11/dist-packages/statsmodels/tsa/base/tsa_model.py:837: ValueWarning: No supported index is available. Prediction results will be given with an integer index beginning at `start`.
  return get_prediction_index(
/usr/local/lib/python3.11/dist-packages/statsmodels/tsa/base/tsa_model.py:837: FutureWarning: No supported index is available. In the next version, calling this method in a model without a supported index will result in an exception.
  return get_prediction_index(


## Summary:

### Data Analysis Key Findings

* A subset of the original DataFrame containing the first 200,000 rows was successfully created.
* The 'timestamp' column was converted to datetime objects and set as the DataFrame index after removing the initial header row.
* The 'close' column was extracted into a pandas Series for time series modeling.
* An ARIMA model with order (5, 1, 0) was successfully fitted to the 'close' price series, which contained 199,999 observations.
* The fitted ARIMA model was used to generate predictions for the next 10 time steps, resulting in 10 predicted values.

### Insights or Next Steps

* The ARIMA model diagnostics (Jarque-Bera and Heteroskedasticity tests) indicate that the residuals are neither normally distributed nor have constant variance, suggesting potential areas for model improvement.
* The predictions were generated with an integer index. To make the predictions more interpretable, the next step should involve aligning these predictions with corresponding future timestamps.


In [27]:
from sklearn.metrics import mean_squared_error
import numpy as np

# Get the fitted values (predictions on the training data)
fitted_values = model_fit.fittedvalues

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(close_series, fitted_values))
print(f"RMSE on the training data: {rmse}")

RMSE on the training data: 1.706069760101764
